clippy_lints/src/borrow_deref_ref.rs

```diff
@@ -0,0 +1,127 @@
+    /// let a: &String = s;
+    /// foo(&**s);
+    ///
+    /// fn foo(_: &str){ }
+    /// ```
+    #[clippy::version = "1.59.0"]
+    pub BORROW_DEREF_REF,
+    complexity,
+    "deref on an immutable reference returns the same type as itself"
+}
+
+declare_lint_pass!(BorrowDerefRef => [BORROW_DEREF_REF]);
+
+impl LateLintPass<'_> for BorrowDerefRef {
+    fn check_expr(&mut self, cx: &LateContext<'tcx>, e: &'tcx rustc_hir::Expr<'_>) {
+        // e = & addrof_target = & * deref_target
+        if_chain! {
+            if !e.span.from_expansion();
+            if let ExprKind::AddrOf(_, Mutability::Not, addrof_target) = e.kind;
+            if !addrof_target.span.from_expansion();
+            if let ExprKind::Unary(UnOp::Deref, deref_target) = addrof_target.kind;
+            if !deref_target.span.from_expansion();
+            if !matches!(deref_target.kind, ExprKind::Unary(UnOp::Deref, ..) );
+            let ref_ty = cx.typeck_results().expr_ty(deref_target);
+            if let ty::Ref(_, inner_ty, Mutability::Not) = ref_ty.kind();
+            then{
+
+                if let Some(parent_expr) = get_parent_expr(cx, e) {
+                    // `*(&*deref_target)` should not trigger because `DEREF_ADDROF` will be triggered as well
+                    if matches!(parent_expr.kind, ExprKind::Unary(UnOp::Deref, ..))
+                        && !is_lint_allowed(cx, DEREF_ADDROF, parent_expr.hir_id) {
+                        return;
+                    }
+
+                    // `&mut e` should trigger when `e` looks like :
+                    // 1. foo()
+                    //
+                    // `&mut e` should not trigger when `e` looks like :
+                    // 1. variable: `x`
+                    // 2. field: `x.a`
+                    //
+                    if matches!(parent_expr.kind, ExprKind::AddrOf(_, Mutability::Mut, _))
+                        && matches!(deref_target.kind, ExprKind::Path(..) | ExprKind::Field(..)) {
+                        return;
+                    }
+                }
+
+                span_lint_and_then(
+                    cx,
+                    BORROW_DEREF_REF,
+                    e.span,
+                    "deref on an immutable reference",
+                    |diag| {
+                        diag.span_suggestion(
+                            e.span,
+                            "if you want to reborrow, use",
+                            snippet_opt(cx, deref_target.span).unwrap(),
+                            Applicability::MachineApplicable
+                        );
+
+                        // has deref trait -> give 2 suggestions
+                        // doesn't have deref trait -> give 1 suggestion
+                        match cx
+                            .tcx
+                            .lang_items()
+                            .deref_trait()
+                            .map(|deref_trait_id| implements_trait(cx, inner_ty, deref_trait_id, &[]) )
```

Comment:
Should not suggest by default

```suggestion
                            .map_or(false, |deref_trait_id| implements_trait(cx, inner_ty, deref_trait_id, &[]) )
```